presto-hive/src/main/java/com/facebook/presto/hive/ParquetHiveRecordCursor.java

```diff
@@ -1127,223 +1010,45 @@ public void addValueFromDictionary(int dictionaryId)
         @Override
         public void addBoolean(boolean value)
         {
-            try {
-                writeFieldNameIfSet(generator, fieldName);
-                generator.writeBoolean(value);
-                wroteValue = true;
-            }
-            catch (IOException e) {
-                throw Throwables.propagate(e);
-            }
+            BOOLEAN.writeBoolean(builder, value);
+            wroteValue = true;
         }
 
         @Override
         public void addDouble(double value)
         {
-            try {
-                writeFieldNameIfSet(generator, fieldName);
-                generator.writeNumber(value);
-                wroteValue = true;
-            }
-            catch (IOException e) {
-                throw Throwables.propagate(e);
-            }
+            DOUBLE.writeDouble(builder, value);
+            wroteValue = true;
         }
 
         @Override
         public void addLong(long value)
         {
-            try {
-                writeFieldNameIfSet(generator, fieldName);
-                generator.writeNumber(value);
-                wroteValue = true;
-            }
-            catch (IOException e) {
-                throw Throwables.propagate(e);
-            }
+            BIGINT.writeLong(builder, value);
+            wroteValue = true;
         }
 
         @Override
         public void addBinary(Binary value)
         {
-            try {
-                writeFieldNameIfSet(generator, fieldName);
-                // todo don't assume binary is a utf-8 string
-                byte[] bytes = value.getBytes();
-                generator.writeUTF8String(value.getBytes(), 0, bytes.length);
-                wroteValue = true;
-            }
-            catch (IOException e) {
-                throw Throwables.propagate(e);
-            }
+            // todo don't assume binary is a utf-8 string
+            byte[] bytes = value.getBytes();
+            VARCHAR.writeSlice(builder, Slices.wrappedBuffer(bytes));
```

Comment:
I would agree. This is much cleaner way to pass complex type involved Expression to workers. I've removed `SliceLiteral` and removed the dirty hack.